File: mario-rust/src/game/world.rs
use macroquad::file::load_string;
use macroquad::prelude::*;

use super::{physics, Config};

const FALLBACK_LEVEL: &str = "\
................................\n\
................................\n\
................................\n\
................................\n\
.......C.........C.......C......\n\
......#####.....#####...#####...\n\
..P....M....E................G..\n\
#######...########..######...###\n";

pub struct World {
    pub solids: Vec<Rect>,
    solid_tiles: Vec<bool>,
    pub coins: Vec<Vec2>,
    pub mushrooms: Vec<Vec2>,
    pub enemy_spawns: Vec<Vec2>,
    pub player_spawn: Vec2,
    pub goal_tile: Vec2,
    pub width: usize,
    pub height: usize,
}

impl World {
    pub async fn load(path: &str, config: &Config) -> Self {
        match load_string(path).await {
            Ok(contents) => match Self::from_ascii(&contents, config) {
                Ok(world) => world,
                Err(error) => {
                    eprintln!("Level parse error: {error}. Using fallback level.");
                    Self::from_ascii(FALLBACK_LEVEL, config).expect("Fallback level is invalid")
                }
            },
            Err(error) => {
                eprintln!("Level load error: {error}. Using fallback level.");
                Self::from_ascii(FALLBACK_LEVEL, config).expect("Fallback level is invalid")
            }
        }
    }

    pub fn from_ascii(contents: &str, config: &Config) -> Result<Self, String> {
        let lines: Vec<&str> = contents
            .lines()
            .map(str::trim_end)
            .filter(|line| !line.is_empty())
            .collect();

        let height = lines.len();
        let width = lines
            .iter()
            .map(|line| line.chars().count())
            .max()
            .unwrap_or(0);

        if width == 0 || height == 0 {
            return Err("Level has no tiles".to_string());
        }

        let tile_size = config.tile_size;
        let mut solid_tiles = vec![false; width * height];
        let mut solids = Vec::new();
        let mut coins = Vec::new();
        let mut mushroom_tiles = Vec::new();
        let mut enemy_spawns = Vec::new();
        let mut player_spawn = None;
        let mut goal_tile = None;

        for (row, line) in lines.iter().enumerate() {
            for (col, ch) in line.chars().enumerate() {
                let world_x = col as f32 * tile_size;
                let world_y = row as f32 * tile_size;
                let tile_pos = vec2(world_x, world_y);

                match ch {
                    '#' => {
                        solid_tiles[row * width + col] = true;
                        solids.push(physics::rect_at(tile_pos, vec2(tile_size, tile_size)));
                    }
                    'C' => coins.push(vec2(world_x + tile_size * 0.5, world_y + tile_size * 0.5)),
                    'M' => mushroom_tiles.push(tile_pos),
                    'E' => enemy_spawns.push(tile_pos),
                    'P' => {
                        if player_spawn.is_some() {
                            return Err("Multiple player spawns found".to_string());
                        }
                        player_spawn = Some(tile_pos);
                    }
                    'G' => {
                        if goal_tile.is_some() {
                            return Err("Multiple goal tiles found".to_string());
                        }
                        goal_tile = Some(tile_pos);
                    }
                    '.' => {}
                    _ => {
                        return Err(format!("Unexpected tile '{ch}'"));
                    }
                }
            }
        }

        let player_spawn = player_spawn.ok_or_else(|| "Missing player spawn".to_string())?;
        let goal_tile = goal_tile.ok_or_else(|| "Missing goal tile".to_string())?;

        let mut world = Self {
            solids,
            solid_tiles,
            coins,
            mushrooms: Vec::new(),
            enemy_spawns,
            player_spawn,
            goal_tile,
            width,
            height,
        };

        world.mushrooms = mushroom_tiles
            .into_iter()
            .map(|tile_pos| {
                let size = config.mushroom_size;
                let tile = config.tile_size;
                let x = tile_pos.x + (tile - size.x) * 0.5;
                let sample_x = tile_pos.x + tile * 0.5;
                let base_y = world
                    .ground_y_for_x(sample_x, tile_pos.y, config)
                    .unwrap_or(tile_pos.y + tile);
                let y = base_y - size.y;
                vec2(x, y)
            })
            .collect();

        Ok(world)
    }

    pub fn draw(&self, config: &Config) {
        let tile = config.tile_size;

        for solid in &self.solids {
            draw_rectangle(
                solid.x,
                solid.y,
                solid.w,
                solid.h,
                Color::new(0.25, 0.55, 0.25, 1.0),
            );
        }

        for coin in &self.coins {
            draw_circle(coin.x, coin.y, tile * 0.2, Color::new(0.95, 0.8, 0.2, 1.0));
        }

        self.draw_mushrooms(config);
        self.draw_goal_post(config);
    }

    pub fn goal_trigger_rect(&self, config: &Config) -> Rect {
        let tile = config.tile_size;
        let goal_center_x = self.goal_tile.x + tile * 0.5;
        let base_y = self
            .ground_y_for_x(goal_center_x, self.goal_tile.y, config)
            .unwrap_or(self.goal_tile.y + tile);

        let pole_height = tile * 3.0;
        let pole_w = tile * 0.18;
        let pole_x = goal_center_x - pole_w * 0.5;
        let pole_y = base_y - pole_height;

        Rect::new(pole_x, pole_y, pole_w, pole_height)
    }

    fn draw_mushrooms(&self, config: &Config) {
        let size = config.mushroom_size;
        for pos in &self.mushrooms {
            let stem_w = size.x * 0.35;
            let stem_h = size.y * 0.45;
            let stem_x = pos.x + (size.x - stem_w) * 0.5;
            let stem_y = pos.y + size.y - stem_h;
            draw_rectangle(
                stem_x,
                stem_y,
                stem_w,
                stem_h,
                Color::new(0.95, 0.9, 0.75, 1.0),
            );

            let cap_h = size.y * 0.6;
            draw_rectangle(
                pos.x,
                pos.y,
                size.x,
                cap_h,
                Color::new(0.85, 0.15, 0.55, 1.0),
            );
            draw_rectangle(
                pos.x + size.x * 0.15,
                pos.y + cap_h * 0.25,
                size.x * 0.2,
                cap_h * 0.35,
                WHITE,
            );
        }
    }

    pub fn is_solid_tile(&self, col: i32, row: i32) -> bool {
        if col < 0 || row < 0 {
            return false;
        }
        let col = col as usize;
        let row = row as usize;
        if col >= self.width || row >= self.height {
            return false;
        }
        self.solid_tiles[row * self.width + col]
    }

    pub fn ground_y_for_x(&self, world_x: f32, start_y: f32, config: &Config) -> Option<f32> {
        let tile = config.tile_size;
        let col = (world_x / tile).floor() as i32;
        let start_row = (start_y / tile).floor().max(0.0) as i32;
        for row in start_row..(self.height as i32) {
            if self.is_solid_tile(col, row) {
                return Some(row as f32 * tile);
            }
        }
        None
    }

    pub fn camera_for_focus(&self, focus: Vec2, config: &Config) -> Camera2D {
        let world_w = self.width as f32 * config.tile_size;
        let world_h = self.height as f32 * config.tile_size;
        let screen_w = screen_width();
        let screen_h = screen_height();

        let mut cam_x = focus.x;
        let mut cam_y = focus.y;

        if world_w > screen_w {
            cam_x = cam_x.clamp(screen_w * 0.5, world_w - screen_w * 0.5);
        } else {
            cam_x = world_w * 0.5;
        }

        if world_h > screen_h {
            cam_y = cam_y.clamp(screen_h * 0.5, world_h - screen_h * 0.5);
        } else {
            cam_y = world_h * 0.5;
        }

        Camera2D {
            target: vec2(cam_x, cam_y),
            zoom: vec2(2.0 / screen_w, 2.0 / screen_h),
            ..Default::default()
        }
    }

    fn draw_goal_post(&self, config: &Config) {
        let tile = config.tile_size;
        let goal_center_x = self.goal_tile.x + tile * 0.5;
        let base_y = self
            .ground_y_for_x(goal_center_x, self.goal_tile.y, config)
            .unwrap_or(self.goal_tile.y + tile);

        let pole_height = tile * 3.0;
        let pole_w = tile * 0.12;
        let pole_x = self.goal_tile.x + tile * 0.5 - pole_w * 0.5;
        let pole_y = base_y - pole_height;

        draw_rectangle(pole_x, pole_y, pole_w, pole_height, GRAY);
        draw_rectangle(
            pole_x + pole_w,
            pole_y + tile * 0.3,
            tile * 0.55,
            tile * 0.35,
            RED,
        );
        draw_rectangle(
            self.goal_tile.x + tile * 0.4,
            base_y - tile * 0.12,
            tile * 0.2,
            tile * 0.12,
            DARKBROWN,
        );
    }
}

File: mario-cpp/core/src/world.cpp
#include "mario/core/world.hpp"

#include <algorithm>
#include <cctype>
#include <optional>
#include <string_view>

#include "mario/core/types.hpp"

namespace mario::core {
namespace {

std::string_view trim_end(std::string_view s) {
  while (!s.empty()) {
    const unsigned char ch = static_cast<unsigned char>(s.back());
    if (ch == '\r' || ch == ' ' || ch == '\t' || ch == '\v' || ch == '\f') {
      s.remove_suffix(1);
      continue;
    }
    break;
  }
  return s;
}

}  // namespace

bool World::from_ascii(std::string_view contents, const Config& config, World& out, std::string& error) {
  std::vector<std::string_view> lines;
  lines.reserve(64);

  std::size_t pos = 0;
  while (pos <= contents.size()) {
    const std::size_t next_nl = contents.find('\n', pos);
    const std::size_t end = (next_nl == std::string_view::npos) ? contents.size() : next_nl;
    std::string_view line = contents.substr(pos, end - pos);
    line = trim_end(line);
    if (!line.empty()) {
      lines.push_back(line);
    }

    if (next_nl == std::string_view::npos) {
      break;
    }
    pos = next_nl + 1;
  }

  const int height = static_cast<int>(lines.size());
  int width = 0;
  for (const auto line : lines) {
    width = std::max(width, static_cast<int>(line.size()));
  }

  if (width <= 0 || height <= 0) {
    error = "Level has no tiles";
    return false;
  }

  out = World{};
  out.width = width;
  out.height = height;

  out.solid_tiles.assign(static_cast<std::size_t>(width * height), 0);
  out.solids.clear();
  out.coins.clear();
  out.mushrooms.clear();
  out.enemy_spawns.clear();

  std::vector<Vec2> mushroom_tiles;
  std::optional<Vec2> player_spawn;
  std::optional<Vec2> goal_tile;

  const units_t tile = config.tile_size;

  for (int row = 0; row < height; ++row) {
    const std::string_view line = lines[static_cast<std::size_t>(row)];
    for (int col = 0; col < static_cast<int>(line.size()); ++col) {
      const char ch = line[static_cast<std::size_t>(col)];
      const units_t world_x = static_cast<units_t>(col) * tile;
      const units_t world_y = static_cast<units_t>(row) * tile;
      const Vec2 tile_pos{world_x, world_y};

      switch (ch) {
        case '#': {
          out.solid_tiles[static_cast<std::size_t>(row * width + col)] = 1;
          out.solids.push_back(Rect{tile_pos.x, tile_pos.y, tile, tile});
        } break;
        case 'C':
          out.coins.push_back(Vec2{world_x + tile / 2, world_y + tile / 2});
          break;
        case 'M':
          mushroom_tiles.push_back(tile_pos);
          break;
        case 'E':
          out.enemy_spawns.push_back(tile_pos);
          break;
        case 'P':
          if (player_spawn.has_value()) {
            error = "Multiple player spawns found";
            return false;
          }
          player_spawn = tile_pos;
          break;
        case 'G':
          if (goal_tile.has_value()) {
            error = "Multiple goal tiles found";
            return false;
          }
          goal_tile = tile_pos;
          break;
        case '.':
          break;
        default:
          error = std::string("Unexpected tile '") + ch + "'";
          return false;
      }
    }
  }

  if (!player_spawn.has_value()) {
    error = "Missing player spawn";
    return false;
  }
  if (!goal_tile.has_value()) {
    error = "Missing goal tile";
    return false;
  }

  out.player_spawn = *player_spawn;
  out.goal_tile = *goal_tile;

  out.mushrooms.reserve(mushroom_tiles.size());
  for (const Vec2 tile_pos : mushroom_tiles) {
    const Vec2 size = config.mushroom_size;
    const units_t x = tile_pos.x + (tile - size.x) / 2;
    const units_t sample_x = tile_pos.x + tile / 2;
    const units_t base_y =
        out.ground_y_for_x(sample_x, tile_pos.y, config).value_or(tile_pos.y + tile);
    const units_t y = base_y - size.y;
    out.mushrooms.push_back(Vec2{x, y});
  }

  return true;
}

bool World::is_solid_tile(int col, int row) const {
  if (col < 0 || row < 0) {
    return false;
  }
  if (col >= width || row >= height) {
    return false;
  }
  return solid_tiles[static_cast<std::size_t>(row * width + col)] != 0;
}

std::optional<units_t> World::ground_y_for_x(units_t world_x, units_t start_y,
                                            const Config& config) const {
  const units_t tile = config.tile_size;
  const int col = static_cast<int>(floor_div(world_x, tile));
  const int start_row = std::max<int>(0, static_cast<int>(floor_div(start_y, tile)));

  for (int row = start_row; row < height; ++row) {
    if (is_solid_tile(col, row)) {
      return static_cast<units_t>(row) * tile;
    }
  }
  return std::nullopt;
}

Rect World::goal_trigger_rect(const Config& config) const {
  const units_t tile = config.tile_size;
  const units_t goal_center_x = goal_tile.x + tile / 2;
  const units_t base_y = ground_y_for_x(goal_center_x, goal_tile.y, config).value_or(goal_tile.y + tile);

  const units_t pole_height = tile * 3;
  const units_t pole_w = (tile * 9) / 50;  // tile * 0.18
  const units_t pole_x = goal_center_x - pole_w / 2;
  const units_t pole_y = base_y - pole_height;
  return Rect{pole_x, pole_y, pole_w, pole_height};
}

}  // namespace mario::core

